src/z3c/pt/tests/test_expressions.py

```diff
@@ -0,0 +1,150 @@
+
+    def test_translate_invalid_path(self):
+        from chameleon.exc import ExpressionError
+
+        expr = expressions.PathExpr('')
+        with self.assertRaises(ExpressionError):
+            expr.translate('not valid', None)
+
+    def test_translate_components(self):
+        from chameleon.codegen import TemplateCodeGenerator
+        from chameleon.astutil import ASTCodeGenerator
+        from chameleon.astutil import node_annotations
+        expr = expressions.PathExpr('')
+        comps = expr._find_translation_components(['a'])
+
+        # First component is skipped
+        self.assertEqual(comps, [])
+
+        # Single literal
+        comps = expr._find_translation_components(['a', 'b'])
+        self.assertEqual([s.s for s in comps], ['b'])
+
+        # Multiple literals
+        comps = expr._find_translation_components(['a', 'b', 'c'])
+        self.assertEqual([s.s for s in comps], ['b', 'c'])
+
+        # Single interpolation---must be longer than one character
+        comps = expr._find_translation_components(['a', '?b'])
+        self.assertEqual([s.s for s in comps], ['?b'])
+
+        comps = expr._find_translation_components(['a', '?var'])
+        self.assertEqual(len(comps), 1)
+        code = ASTCodeGenerator(comps[0]).code
+        self.assertEqual(code, '(format % args)')
+        code = TemplateCodeGenerator(comps[0]).code
+        self.assertEqual(code, "('%s' % (var, ))")
+
+        args = node_annotations[comps[0].right]
+        code = TemplateCodeGenerator(args).code
+        self.assertEqual(code, '(var, )')
+
+
+        # Multiple interpolations
+        comps = expr._find_translation_components(['a', '?var', "?var2"])
+        self.assertEqual(len(comps), 2)
+        code = ASTCodeGenerator(comps[0]).code
+        self.assertEqual(code, '(format % args)')
+        code = TemplateCodeGenerator(comps[0]).code
+        self.assertEqual(code, "('%s' % (var, ))")
+
+        args = node_annotations[comps[0].right]
+        code = TemplateCodeGenerator(args).code
+        self.assertEqual(code, '(var, )')
+
+        code = ASTCodeGenerator(comps[1]).code
+        self.assertEqual(code, '(format % args)')
+        code = TemplateCodeGenerator(comps[1]).code
+        self.assertEqual(code, "('%s' % (var2, ))")
+
+        args = node_annotations[comps[1].right]
+        code = TemplateCodeGenerator(args).code
+        self.assertEqual(code, '(var2, )')
+
+        translated = expr.translate("a/?var/?var2", None)
+        self.assertEqual(len(translated), 1)
+        code = TemplateCodeGenerator(translated[0]).code
+        # XXX: Normally this starts with 'None =', but sometimes on Python 2,
+        # at least in tox, it starts with '__package__ ='. Why
+        # is this?
+        code = code.strip().replace("__package__", 'None')
+        self.assertEqual(
+            code,
+            "None = _path_traverse(a, econtext, True, (('%s' % (var, )), ('%s' % (var2, )), ))")
```

Comment:
Is that even legal code?  You can't assign to `None`, you get a `SyntaxError: can't assign to keyword`.

Reply:
Yah, it's not legal. It's just because of how I'm calling this particular method (passing a target of None), which I assume the real code wouldn't do.

Reply:
I'm not a fan of unit tests that do things no real code would ever do ;)

Although here TBH I'm more curious about where the `__package__` comes from.